### DISCUSSION TITLE
Problems with the folder picker - what am I doing wrong?

### DISCUSSION THREAD

**Discussion by c64audio** [AvaloniaUI/Avalonia#11561] [Q&A] [answered]
I've got a problem with the OpenFolderPickerAsync command on Windows and Mac (using 11 preview 8, behaved the same on 11 preview 1).

Trying to do what should be a simple thing: throw up a folder picker in the middle of some code.

The code is below. I've tried invoking the picker in two ways: once in the normal thread, once in the UI thread.

On Windows, using the normal method works as expected (yay), but when you use the UI thread, you never see a picker.

On Mac, the normal thread crashes with a nasty NSException, and if you use the UI thread, you never see a picker (changing the dispatcherpriority doesn't seem to help).

```
        public string SelectFolder(string title)
        {
            Task<string> task = Task.Run(() => SelectFolderAsync(title));
            return task.Result;
        }

        public async Task<string> SelectFolderAsync(string title)
        {
            FolderPickerOpenOptions opts = new()
            {
                Title = title,
                AllowMultiple = false
            };

// ui thread - this doesn't show a picker on Windows nor Mac
            IReadOnlyList<IStorageFolder> folderPaths = await Dispatcher.UIThread.InvokeAsync(() => this.StorageProvider.OpenFolderPickerAsync(opts), DispatcherPriority.Input);

// normal thread - works on Windows, crashes on Mac 10.15
            //IReadOnlyList<IStorageFolder> folderPaths = await this.StorageProvider.OpenFolderPickerAsync(opts);

            if (folderPaths.Count > 0)
            {
                return folderPaths[0].Name;
            }
            return string.Empty;
        }

```

Here's what the Mac crash looks like when you're not using the UI thread:

  ???                                 0x0000000124867cd3 0x0 + 4907760851
        19  ???                                 0x0000000121d94b9b 0x0 + 4862856091
        20  ???                                 0x0000000121d97f92 0x0 + 4862869394
        21  ???                                 0x000000012486720d 0x0 + 4907758093
        22  ???                                 0x0000000121d19651 0x0 + 4862350929
        23  ???                                 0x0000000121b7d716 0x0 + 4860663574
        24  ???                                 0x0000000121b63ee6 0x0 + 4860559078
        25  ???                                 0x0000000121b78908 0x0 + 4860643592
        26  ???                                 0x0000000121b78818 0x0 + 4860643352
        27  ???                                 0x0000000121b787b7 0x0 + 4860643255
        28  ???                                 0x0000000121b741dc 0x0 + 4860625372
        29  ???                                 0x0000000121b5880e 0x0 + 4860512270
        30  libcoreclr.dylib                    0x000000010b79b5d9 CallDescrWorkerInternal + 124
        31  libcoreclr.dylib                    0x000000010b5e730f _Z18DispatchCallSimplePmjmj + 223
        32  libcoreclr.dylib                    0x000000010b5fe488 _ZN12ThreadNative20KickOffThread_WorkerEPv + 136
        33  libcoreclr.dylib                    0x000000010b5af808 _ZL31ManagedThreadBase_DispatchOuterP22ManagedThreadCallState + 296
        34  libcoreclr.dylib                    0x000000010b5afd70 _ZN17ManagedThreadBase7KickOffEPFvPvES0_ + 32
        35  libcoreclr.dylib                    0x000000010b5fe55a _ZN12ThreadNative13KickOffThreadEPv + 170
        36  libcoreclr.dylib                    0x000000010b4b13f7 _ZN7CorUnix10CPalThread11ThreadEntryEPv + 407
        37  libsystem_pthread.dylib             0x00007fff6f4b7109 _pthread_start + 148
        38  libsystem_pthread.dylib             0x00007fff6f4b2b8b thread_start + 15
)
libc++abi.dylib: terminating with uncaught exception of type NSException

**Comment by maxkatz6** [comment, marked as answer]
> what am I doing wrong? 

This:
```
        public string SelectFolder(string title)
        {
            Task<string> task = Task.Run(() => SelectFolderAsync(title));
            return task.Result;
        }
```

You **should not** do sync over async in GUI applications.

> // ui thread - this doesn't show a picker on Windows nor Mac

Because you blocked UI thread by SelectFolder call above.

> // normal thread - works on Windows, crashes on Mac 10.15

Because macOS requires file picker to run on UI thread.

In 11.0 there are some hacks to make it work, but these only help with macos, but not mobile/wasm.

Make your code properly async. And it will work as designed - asynchronously and on UI thread.

**Comment by c64audio** [comment]
I accept that this is the answer (thanks), but I'm going to need the hack. Let me explain why.

For context, this is an install script, which is triggered from a form button. 

The script obviously does a number of things in a strict order, and doesn't require the calling form/UI to be active or responsive during that process. It also doesn't have to run on IOS or Android.

So, this install script (unfortunately) requires the user to be asked for a folder path an arbitrary number of times: this means I can't gather this information into a form field before the install button is pressed. In this particular use case, a modal folder dialog box would have been an ideal solution. But the Gods seem to have decided that my use case is morally corrupt or something.

So, if there's a hack here for MacOSX to make this process sane, I'll happily take it: re-factoring entirely satisfactory working code into an asynchronous mess just so I can get this one item of data from the user seems to be the tail wagging the dog. 

Chris

**Comment by c64audio** [comment]
Ho hum, well, I solved the problem with the minimal amount of faffing by separating out a bit of logic and putting that on a worker thread so I could dispatch the folder request back to the UI thread when necessary.
